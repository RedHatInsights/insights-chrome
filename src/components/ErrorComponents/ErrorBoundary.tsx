import React from 'react';

import ErrorComponent from './DefaultErrorComponent';
import AccountOnHoldError, { checkAccountOnHold } from './AccountOnHoldError';

type ErrorBoundaryState = {
  hasError: boolean;
  error?: any;
  errorInfo?: any;
};

const INVALID_AUTH_STATE_ERROR = 'No matching state found in storage';

class ErrorBoundary extends React.Component<
  {
    children: React.ReactNode;
    singIn?: () => Promise<void>;
  },
  ErrorBoundaryState
> {
  state: ErrorBoundaryState = {
    hasError: false,
  };

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  handleInvalidAuthState(): void {
    const repairedUrl = new URL(window.location.href);
    // remove invalid SSO state and force re authentication
    repairedUrl.hash = '';
    // remove possibly broken local storage state from client
    localStorage.clear();
    // hard page reload
    window.location.href = repairedUrl.toString();
  }

  componentDidCatch(error: any, errorInfo: any) {
    console.error('Chrome encountered an error!', error);
    this.setState((prev) => ({
      ...prev,
      error,
      errorInfo,
    }));

    if (typeof error.message === 'string' && error.message === INVALID_AUTH_STATE_ERROR) {
      this.handleInvalidAuthState();
    }
  }

  render() {
    if (this.state.hasError) {
      if (checkAccountOnHold(this.state.error)) {
        return <AccountOnHoldError error={this.state.error} />;
      }
      return <ErrorComponent error={this.state.error} errorInfo={this.state.errorInfo} signIn={this.props.singIn} />;
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
