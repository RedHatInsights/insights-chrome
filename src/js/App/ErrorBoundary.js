import React from 'react';
import PropTypes from 'prop-types';
import * as Sentry from '@sentry/browser';

class ErrorBoundary extends React.Component {
  state = {
    hasError: false,
  };

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo) {
    console.error('Chrome encountered an error!', error);
    this.setState(() => ({
      errorId: Sentry.captureException(new Error('Support case created'), {
        error,
        errorInfo,
      }),
    }));
  }

  render() {
    if (this.state.hasError) {
      return <h1>Something went wrong. Error ID: {this.state.errorId}</h1>;
    }

    return this.props.children;
  }
}

ErrorBoundary.propTypes = {
  children: PropTypes.node,
};

export default ErrorBoundary;
