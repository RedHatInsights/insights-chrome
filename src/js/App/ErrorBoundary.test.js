/* eslint-disable react/prop-types */
import React from 'react';
import { render } from '@testing-library/react';
jest.mock('@sentry/browser', () => ({
  __esModule: true,
  captureException: () => 'mock-id',
}));

import ErrorBoundary from './ErrorBoundary';

const Throw = () => {
  new Error('Expected error');
};
const DummyComponent = ({ shouldThrow = false }) => <ErrorBoundary>{shouldThrow ? <Throw /> : <button></button>}</ErrorBoundary>;
describe('ErrorBoundary', () => {
  test('should render component with no error', () => {
    const { container } = render(<DummyComponent />);
    const button = container.querySelectorAll('button');
    expect(button).toHaveLength(1);
  });

  test('should render error boundary', () => {
    /**
     * we don't want the error stack here because we expect it
     */
    const spy = jest.spyOn(console, 'error');
    spy.mockImplementation(() => {});
    const { container } = render(<DummyComponent shouldThrow />);
    expect(container).toMatchSnapshot();
    spy.mockRestore();
  });
});
