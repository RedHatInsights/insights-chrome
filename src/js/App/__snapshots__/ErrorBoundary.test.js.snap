// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should render error boundary 1`] = `
<div>
  <h1>
    Something went wrong. Error ID: 
    mock-id
  </h1>
</div>
`;
